### ISSUE TITLE
Language pack

### ISSUE THREAD

**Issue by Demiurge67** [stepmania/stepmania#1838]
How can I Russify the program? I need a Russian language pack. Ready to participate in the work.

**Comment by dguzek** [comment]
StepMania's *_fallback* theme does not have support for Russian at this time.

If you are interested in translating StepMania into Russian, I can try to explain how you could help do that.

**Comment by eintw1ck** [comment]
@Demiurge67 I assume you can get started by translating everything in this file, https://github.com/stepmania/stepmania/blob/master/Themes/_fallback/Languages/en.ini

**Comment by dguzek** [comment]
Yes, that is step 1! 🙂 

@Demiurge67, you'd start by looking at the files in here: https://github.com/stepmania/stepmania/tree/5_1-new/Themes/_fallback/Languages

You would first create a new file `ru.ini` in that folder that follows the format of the existing files.

The other important step would be to create a bitmap sheet using a open source or public domain Cyrillic font.   The *_fallback* theme does not have any Cyrillic support right now, so you would need to add that.

You can start to see how font sheets are laid out by looking at the *_fallback* theme's _roboto black bold [spritesheet](https://github.com/stepmania/stepmania/blob/5_1-new/Themes/_fallback/Fonts/_roboto%20black%20Bold%2024px%20%5Bmain%5D%2015x15.png) and [ini file](https://github.com/stepmania/stepmania/blob/5_1-new/Themes/_fallback/Fonts/_roboto%20black%20Bold%2024px.ini).  Release versions of StepMania for Windows OS come with **./Program/Texture Font Generator.exe** but I am unsure if it supports exporting Cyrillic characters.  You'll have to try it and find out.

**Comment by Demiurge67** [comment]
> 
> 
> StepMania's __fallback_ theme does not have support for Russian at this time.
> 
> If you are interested in translating StepMania into Russian, I can try to explain how you could help do that.

Yes, of course I'll do my best.

**Comment by Demiurge67** [comment]
> 
> 
> Yes, that is step 1! 🙂
> 
> @Demiurge67, you'd start by looking at the files in here: https://github.com/stepmania/stepmania/tree/5_1-new/Themes/_fallback/Languages
> 
> You would first create a new file `ru.ini` in that folder that follows the format of the existing files.
> 
> The other important step would be to create a bitmap sheet using an open source or public domain Cyrillic font. The __fallback_ theme does not have any Cyrillic support right now, so you would need to add that.
> 
> You can start to see how font sheets are laid out by looking at the __fallback_ theme's _roboto black bold [spritesheet](https://github.com/stepmania/stepmania/blob/5_1-new/Themes/_fallback/Fonts/_roboto%20black%20Bold%2024px%20%5Bmain%5D%2015x15.png) and [ini file](https://github.com/stepmania/stepmania/blob/5_1-new/Themes/_fallback/Fonts/_roboto%20black%20Bold%2024px.ini). Release versions of StepMania for Windows OS come with **./Program/Texture Font Generator.exe** but I am unsure if it supports exporting Cyrillic characters. You'll have to try it and find out.

Good. I'll try to get started.

**Comment by dguzek** [comment]
Sounds good.  Let us know how it goes.

**Comment by Demiurge67** [comment]
Some of the sentences are incomprehensible to me. My English is not very good, apparently. Also, there may be specific terminology and contextual concepts that I can't translate. However, maybe in a few days, I will give the results to the distinguished community.

**Comment by Demiurge67** [comment]
You can test my work. 

![_game chars 16px 9x1](https://user-images.githubusercontent.com/51953496/60387871-8489f280-9ab2-11e9-9bf2-5b2dfc91732b.png)

![rus_512_16 16x16](https://user-images.githubusercontent.com/51953496/60387872-85228900-9ab2-11e9-96c7-338bcad7aa04.png)

(url)

**Comment by Demiurge67** [comment]
[_16px fonts.ini.txt](https://github.com/stepmania/stepmania/files/3342174/_16px.fonts.ini.txt)
[rus_512_16 16x16.ini.txt](https://github.com/stepmania/stepmania/files/3342175/rus_512_16.16x16.ini.txt)
[ru.ini.txt](https://github.com/stepmania/stepmania/files/3342177/ru.ini.txt)

**Comment by Demiurge67** [comment]
Hi, friend! How I must edit:
**_16px font.ini
rus_512_16 16x16.ini**
and 
**ru.ini**
files?
Now my font work, but letter too small.

**Comment by dguzek** [comment]
Thank you for working on this.

How did you create the png of the font ([three posts above](https://github.com/stepmania/stepmania/issues/1838#issuecomment-506976632) this one)?  Did you use Texture Font Generator.exe or find and use an existing asset from somewhere else?

It seems that the other Latin-character fonts in the _fallback theme were exported as 24px per glyph.  (_roboto black Bold 24px, _open sans semibold 24px, frutiger 24px).  The png you provided used 16px per glyph.  

If you re-export the font to use 24px per glyph, that may fix the size.

Also, the png you provided has a permanently embedded black stroke around each glyph.  The other fonts in the _fallback and default theme ensure that each glyph is only `#ffffff` so that themes can apply color as they want.  (Some fonts use a secondary png with a `#ffffff` stroke, so that it can be dynamically overlaid in-game.)  If you are able to re-export the font, please keep the font and the stroke separated.

**Comment by Demiurge67** [comment]
I found a broken translator on (vk.com) and taked png from it. My translation is quite usable, but the letters are too small (sorry for repeating).

**Comment by Demiurge67** [comment]
I need more guidance, sorry. Where you can download Texture Font Generator.exe?

**Comment by dguzek** [comment]
I think _Texture Font Generator.exe_ is packed with release versions of SM5 for Windows OS in the `Program` directory.

**Comment by Demiurge67** [comment]
I try this once. This program not support cyryllic letters.

**Comment by Demiurge67** [comment]
I adjusted the file name with the font rus_512_24px 16x16.png. The letters have become normal size.

**Comment by dguzek** [comment]
Thank you for helping and for figuring out how to fix the problem without guidance.

Is there more work to do here?  If yes, that is okay.  Please take your time.

If you are done, I will try to find time to review your work, but I will be busy and may not respond here for several more weeks.  Maybe someone else can review this.

If you'd like, prepare your changes as a [pull request](https://github.com/stepmania/stepmania/pulls).

**Comment by Demiurge67** [comment]
Yes, indeed, there are still problems.For example, on the music selection screen, there is text overlay. 
In addition, I try to shorten the phrases, to formulate them in the most understandable way for the player. 99% of the gameplay interface and settings menu with hints are translated correctly, in my opinion.
Still present in the text incorrectly translated elements (in the game settings) do not have, in my opinion, of fundamental importance.
Also, on your recommendation, I corrected the _roboto black Bold 24px [main] 15x15 file.png. I do not know how much it helped in the work, because I have not yet understood where the use of the font is written. 
However, I believe that the product can be used. At least - my daughter independently understood settings and dances every day.

Important note. I am working with version 5.0.12 because version 5.1.0 b2 starts without displaying the main menu (paid mode?)

**Comment by Demiurge67** [comment]
New files:
[pack.zip](https://github.com/stepmania/stepmania/files/3350274/pack.zip)

**Comment by Horsey-** [comment]
you can hold f3 and an overlay will show up, from there you keep holding f3 and press the key that corresponds to pay/Freeplay/home mode

**Comment by Demiurge67** [comment]
New version of rus.ini. Correct some errors.
[rus.zip](https://github.com/stepmania/stepmania/files/3364832/rus.zip)

**Comment by sergvoronezh** [comment]
New version stepmania 5.0.12 Russian Language
https://cloud.mail.ru/public/Msk8/SwXPfJvxL
https://youtu.be/kqm_6nhnGyM?t=320
Replaced fonts
with Texture_Font_Generator-Euro V0.3
https://josevarela.xyz/SMArchive/Tools/index.html
Works on version 5.0.12
-----------------------------------------------
StepMania 5.1.0-b2
Font sizes and heights are not displayed correctly
https://i.postimg.cc/JnpyBrBN/image.jpg
https://i.postimg.cc/GhW4DS95/image.jpg
---------------------------------------------------
translated by googol translator
